File: 2699-modify-graph-edge-weights/2699-modify-graph-edge-weights.py
import heapq

class Solution:
    def modifiedGraphEdges(self, n, edges, source, destination, target):
        adjs = [{} for _ in range(n)]

        for edge in edges:
            adjs[edge[0]][edge[1]] = edge[2]
            adjs[edge[1]][edge[0]] = edge[2]

        distTo = [float('inf')] * n
        distTo[source] = 0

        pq = [(0, source)]
        heapq.heapify(pq)

        self.dijkstra(adjs, distTo, pq)

        if distTo[destination] == target:
            return self.fill(edges)
        elif distTo[destination] < target:
            return []

        for edge in edges:
            if edge[2] == -1:
                edge[2] = 1
                adjs[edge[0]][edge[1]] = 1
                adjs[edge[1]][edge[0]] = 1

                pq = [(distTo[edge[0]], edge[0]), (distTo[edge[1]], edge[1])]

                self.dijkstra(adjs, distTo, pq)

                if distTo[destination] == target:
                    return self.fill(edges)
                elif distTo[destination] < target:
                    edge[2] += target - distTo[destination]
                    adjs[edge[0]][edge[1]] = edge[2]
                    adjs[edge[1]][edge[0]] = edge[2]
                    return self.fill(edges)

        return []

    def fill(self, edges):
        for edge in edges:
            if edge[2] == -1:
                edge[2] = int(2e9)
        return edges

    def dijkstra(self, adjs, distTo, pq):
        while pq:
            curr_dist, curr = heapq.heappop(pq)

            for next_node, weight in adjs[curr].items():
                if weight > 0:
                    if distTo[next_node] - weight > distTo[curr]:
                        distTo[next_node] = distTo[curr] + weight
                        heapq.heappush(pq, (distTo[next_node], next_node))

File: 2699-modify-graph-edge-weights/2699-modify-graph-edge-weights.java
class Solution {
    public int[][] modifiedGraphEdges(int n, int[][] edges, int source, int destination, int target) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        boolean[][] editable = new boolean[n][n];
        for (int i = 0; i < n; i++)
            Arrays.fill(editable[i], false);
        int[][] updatedEdge = new int[n][n];
        for (int i = 0; i < n; i++)
            Arrays.fill(updatedEdge[i], Integer.MAX_VALUE / 3);
        for (int[] e : edges)   {
            int u = e[0];
            int v = e[1];
            int w = e[2];
            if (w == -1)    {
                editable[u][v] = true;
                editable[v][u] = true;
            }
            updatedEdge[u][v] = Math.max(1, w);
            updatedEdge[v][u] = Math.max(1, w);
            map.computeIfAbsent(u, x -> new ArrayList<>()).add(v);
            map.computeIfAbsent(v, x -> new ArrayList<>()).add(u);
        }
        
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[1] - b[1]);
        Integer[] dist1 = new Integer[n];
        pq.offer(new int[] {destination, 0});
        
        while (!pq.isEmpty())   {
            int[] cur = pq.poll();
            int u = cur[0];
            int w = cur[1];
            if (dist1[u] != null)
                continue;
            dist1[u] = w;
            
            if (!map.containsKey(u))
                continue;
            
            for (int v : map.get(u))    {
                int nw = updatedEdge[u][v];
                if (dist1[v] != null)
                    continue;
                pq.offer(new int[] {v, w + nw});
            }
        }
        pq.clear();
        Integer[] dist = new Integer[n];
        pq.offer(new int[] {source, 0});
        while (!pq.isEmpty())   {
            int[] cur = pq.poll();
            int u = cur[0];
            int w = cur[1];
            if (dist[u] != null)
                continue;
            dist[u] = w;
            if (u == destination && w != target)
                return new int[0][0];
            if (!map.containsKey(u))
                continue;
            
            for (int v : map.get(u))    {
                int nw = updatedEdge[u][v];
                if (dist[v] != null)
                    continue;
                if (editable[u][v] && dist[u] + nw + dist1[v] < target)    {
                    nw = target - dist1[v] - dist[u];
                    updatedEdge[u][v] = nw;
                    updatedEdge[v][u] = nw;
                }
                pq.offer(new int[] {v, w + nw});
            }
        }
        
        for (int i = 0; i < edges.length; i++)  {
            int a = edges[i][0];
            int b = edges[i][1];
            edges[i][2] = updatedEdge[a][b];
        }
        
        return edges;
    }
}